pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
+
+                let deleting = tl
+                    .delete_progress
+                    .try_lock()
+                    .map(|flow| !flow.is_not_started())
+                    .unwrap_or(true);
+
+                if is_same && is_earlier && !deleting {
+                    Some(tl.clone())
+                } else {
+                    None
+                }
+            })
+            .for_each(|timeline| {
+                // important in this scope: we are holding the Tenant::timelines lock
+                let span = tracing::info_span!("reparent", reparented=%timeline.timeline_id);
+
+                tasks.spawn(
+                    async move {
+                        let res = timeline
+                            .remote_client
+                            .as_ref()
+                            .expect("sibling has to have remote client because we have one")
+                            .schedule_reparenting_and_wait(&new_parent)
+                            .await;
+
+                        match res {
+                            Ok(()) => Some(timeline),
+                            Err(e) => {
+                                // with the use of tenant slot, we no longer expect these.
+                                tracing::warn!("reparenting failed: {e:#}");
+                                None
+                            }
+                        }
+                    }
+                    .instrument(span),
+                );
+            });
+
+        let reparenting_candidates = tasks.len();
+        let mut reparented = Vec::with_capacity(tasks.len());
+
+        while let Some(res) = tasks.join_next().await {
+            match res {
+                Ok(Some(timeline)) => {
+                    tracing::info!(reparented=%timeline.timeline_id, "reparenting done");
+                    reparented.push(timeline.timeline_id);
+                }
+                Ok(None) => {
+                    // lets just ignore this for now. one or all reparented timelines could had
+                    // started deletion, and that is fine.
+                }
+                Err(je) if je.is_cancelled() => unreachable!("not used"),
+                Err(je) if je.is_panic() => {
+                    // ignore; it's better to continue with a single reparenting failing (or even
+                    // all of them) in order to get to the goal state... though, this would be
+                    // exceptional, and we have already done N non-cancellable actions. hard to say
+                    // which requires less manual repair.
+                }
+                Err(je) => tracing::error!("unexpected join error: {je:?}"),
+            }
+        }
+
+        if reparenting_candidates != reparented.len() {
+            tracing::info!("failed to reparent some candidates");
+        }
+
+        rtc.schedule_detaching_from_ancestor_and_wait((ancestor.timeline_id, ancestor_lsn))
+            .await?;
```

Comment:
> Which means, after PS comes back up again, the layers that we uploaded in schedule_adding_existing_layers_to_index_upload_and_wait are subject to garbage collection, e.g., because they're past the `ancestor_lsn`

Or worse than gc: current compaction changes some or all straddling layers or compaction changes on detached timeline. Gc is quite unlikely unless some reparenting failed (because of timeline deletion), because at this point, the reparentings are done.

I guess if we fail at the last step, we should only try reparenting instead of copying the layers. Before that, this cannot be used as retrying (by control plane)...

Reply:
Needs more testing, but not on this PR.

Reply:
Hm, seems like a fundamental flaw in the RFC to me; the RFC should specify how we deal with this very conceivable condition.

If you insist on merging this so the PR doesn't grow too old / large, then let's merge it, but please make sure this ends up in the "todo before release" list for this API, and is discussed in the RFC.

Reply:
Solution approaches:

- block gc and compaction on the ancestor timeline (index_part.json change) until it's retried to completion
    - I was going for this option earlier on, nothing wrong there except that it needs the index_part.json change
    - it's quite ackward to inspect state of a any root timeline's index_part.json to decide to skip compaction+gc, but manageable

Perhaps I should switch the order to make this non-retryable but still safe to test before I get those index_part.json changes.